core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java

```diff
@@ -167,15 +167,21 @@ private void addIndexDependencyPaths(List<IndexDependencyBuildItem> indexDepende
         if (indexDependencyBuildItems.isEmpty()) {
             return;
         }
-        final Set<ArtifactKey> indexDependencyKeys = new HashSet<>(indexDependencyBuildItems.size());
+        final Set<ArtifactKey> indexDependencyKeys = new HashSet<>();
+        final Set<String> indexGroupIds = new HashSet<>();
         for (IndexDependencyBuildItem indexDependencyBuildItem : indexDependencyBuildItems) {
-            indexDependencyKeys.add(ArtifactKey.of(indexDependencyBuildItem.getGroupId(),
-                    indexDependencyBuildItem.getArtifactId(),
-                    indexDependencyBuildItem.getClassifier(),
-                    ArtifactCoords.TYPE_JAR));
+            if (indexDependencyBuildItem.getArtifactId() != null) {
+                indexDependencyKeys.add(ArtifactKey.of(indexDependencyBuildItem.getGroupId(),
+                        indexDependencyBuildItem.getArtifactId(),
+                        indexDependencyBuildItem.getClassifier(),
+                        ArtifactCoords.TYPE_JAR));
+            } else {
+                indexGroupIds.add(indexDependencyBuildItem.getGroupId());
+            }
         }
         for (ResolvedDependency dep : curateOutcomeBuildItem.getApplicationModel().getDependencies()) {
-            if (dep.isRuntimeCp() && indexDependencyKeys.contains(dep.getKey())) {
+            if (dep.isRuntimeCp()
+                    && (indexDependencyKeys.contains(dep.getKey()) || indexGroupIds.contains(dep.getGroupId()))) {
                 for (Path path : dep.getContentTree().getRoots()) {
```

Comment:
@famod i think the problem is here, it does not respect filters. Ideally it should walk `dep.getContentTree()` instead.

Reply:
@aloubyansky could you be more specific? What filters do you have in mind?

Reply:
Some modules produce multiple JARs with different classifiers from the same output directory. They use include/exclude filters to identify content that should be included to each specific JAR and this is what `dep.getContentTree()` is for, i.e. in some cases it might be a subset of what's found under the `root` path.

Reply:
Ah ok, so you're talking about a multi-module project where the dependency is a module directory (not the final jar), right? I'm not sure how to handle this though. If you have an idea pls feel free to add something to this PR. Alternatively, we could create a new issue and address this problem in a separate PR.

Reply:
Right, it'd be out of scope of this change to enhance indexing to support that.